### DISCUSSION TITLE
Build issues on MacOS

### DISCUSSION THREAD

**Discussion by loopless** [ginkgo-project/ginkgo#1643] [General]
I am trying to build on macOS Sonoma 14.5, using the homebrew clang 18.1.4 installation
I configured with the following command

cmake -G "Unix Makefiles" .. -DCMAKE_C_COMPILER=/opt/homebrew/opt/llvm/bin/clang -DCMAKE_CXX_COMPILER=/opt/homebrew/opt/llvm/bin/clang++ 

The CMake configuration step went fine, but when linking the first .dylib

[  7%] Linking CXX shared library ../lib/libginkgo_reference.dylib
Undefined symbols for architecture arm64:
  "gko::batch::preconditioner::Jacobi<std::__1::complex<double>, int>::Jacobi(std::__1::shared_ptr<gko::Executor const>)", referenced from:
      gko::EnablePolymorphicObject<gko::batch::preconditioner::Jacobi<std::__1::complex<double>, int>, gko::batch::BatchLinOp>::create_default_impl(std::__1::shared_ptr<gko::Executor const>) const in batch_bicgstab_kernels.cpp.o
      gko::EnablePolymorphicObject<gko::batch::preconditioner::Jacobi<std::__1::complex<double>, int>, gko::batch::BatchLinOp>::clear_impl() in batch_bicgstab_kernels.cpp.o
  "gko::batch::preconditioner::Jacobi<std::__1::complex<float>, int>::Jacobi(std::__1::shared_ptr<gko::Executor const>)", referenced from:
 <and many more link errors>

Clearly I am missing something quite straightforward here...

**Comment by MarcelKoch** [comment]
Thanks for bringing this up. Which Ginkgo version are you using? Are you building a shared library, or a static one (this can be found in the CMake output for Ginkgo)?

**Reply by upsj** [reply]
This is a shared build. We seem to be using `core` code inside `reference`, not sure why that isn't caught by our circular dependency check.

**Comment by greole** [comment]
As a quick fix you could try to compile with AppleClang, this seems to work for me at least for version 14 and 15.

**Comment by loopless** [comment]
Switching to Apple Clang makes the link errors go away and build succeeds.
 Unfortunately Apple Clang does not support OpenMP.

I am wondering what the issue is with homebrew clang 18.1.4. I have used that compiler for building other projects without issues.
It is odd that the errors are for symbols internal to gingko. I'll look into this a bit more, and do some home-brew updates.

I am attaching the CMake output from the homebrew clang.

[ginkgo_cmake.txt](https://github.com/user-attachments/files/16145133/ginkgo_cmake.txt)

**Reply by Fabian188** [reply]
> Switching to Apple Clang makes the link errors go away and build succeeds. Unfortunately Apple Clang does not support OpenMP.

I use AppleClang with llvm openmp from homebrew without issues.

**Comment by tpadioleau** [comment]
Hi there, a colleague noticed the same kind of error. On our side building static libraries seems to be a workaround. Did you find a better solution ?

**Comment by yhmtsai** [comment]
It sounds similar to https://github.com/ginkgo-project/ginkgo/issues/1654.
We have fixed that in https://github.com/ginkgo-project/ginkgo/pull/1724
which version of ginkgo did you try?

**Reply by tpadioleau** [reply]
We installed the version 1.8, we will try 1.9 as soon as possible then :)

**Reply by tpadioleau** [reply]
Seems to compile fine, thanks

**Comment by yhmtsai** [comment]
As @tpadioleau confirms that it works after #1724, close this discussion.
Please feel free to create an new one if you face any build issue.
